### PR TITLE
[ASM] [HIP] feat(fmha_bwd): add bf16 hd=256 fused dKdV+dQ ASM kernels for gfx950

### PR DESCRIPTION
**Pull request by alexioslyrakis-amd**
## Summary

Add hand-written ASM backward attention kernels for **bf16 hd=256** on **gfx950 (MI355X)**:

- **Fused dKdV+dQ kernel**: Two-phase single-launch kernel. Phase 1 computes dK/dV (Q-walk with K/V resident in registers). Phase 2 computes dQ (K-walk with Q/dO resident). Writes dQ directly as bf16 — no atomic accumulation, no post-processing needed.
- **ODO preprocessing kernel**: Computes D = dO · O per token using coalesced loads and LDS tree reduction. Layout-agnostic (reads head stride from args).

## Supported contract

- **Layout**: group/varlen THD (mode=1), non-causal (mask=0)
- **Precision**: BF16, hd_qk=256, hd_v=256
- **Workgroup**: 512 threads (8 waves) for dQdKdV, 256 threads for ODO
- **Atomic**: a16 (atomic32=0) — no post-processing needed
- **Causal / a32 / dense BSHD**: not supported; fall through to CK

### Gaps vs sibling ASM kernels

Callers relying on a sibling kernel's coverage will silently fall back to CK for these cases:

- **mode=0 (batch/dense)**: not supported. Siblings have it.
- **mask=1/mask=2 (causal)**: not supported. The fp8bf16 hd256 backward already supports causal; this kernel does not yet.
- **atomic32=1 (a32)**: not supported. Only a16 mode.
- **fp16**: not supported. Only bf16.
- **GQA/MQA**: rejected by gate (`nhead_q == nhead_k`). Caller falls back to CK — no wrong-answer risk.

## Changes

| File | Description |
|------|-------------|
| `hsa/gfx950/fmha_v3_bwd/bwd_hd256_bf16_a16_pssk_group.co` | Fused dQdKdV kernel (512 threads, 8 waves) |
| `hsa/gfx950/fmha_v3_bwd/bwd_hd256_odo_bf16_group.co` | ODO preprocessing kernel (256 threads) |
| `hsa/gfx950/fmha_v3_bwd/fmha_bwd_dqdkdv.csv` | +1 entry: bf16, hd256, noncausal, group, pssk=1 |
| `hsa/gfx950/fmha_v3_bwd/fmha_bwd_odo.csv` | +1 entry: bf16, hd256, group mode |
| `csrc/cpp_itfs/mha_bwd.cu` | bdx=512 for hd>=256 |
| `aiter/ops/mha.py` | Enable gfx950 bwd dispatch for hd256: hdim gate, atomic32=0, causal guard |

## Performance

Tested on MI355X (gfx950) using `bwd.exe` with `AITER_ASM_DIR=<repo>/hsa`:

```
./bwd.exe -prec=bf16 -b=16 -h=18 -d=256 -s=2560 -mask=0 -v3_atomic_fp32=0 -mode=1 -kname=1 -v=0
```

Note: stock `generate_seqlens()` randomizes segment lengths in group mode (total tokens preserved). Numbers below are with randomized segments.

| Configuration | CK (baseline) | ASM (this PR) | Speedup |
|--------------|---------------|---------------|---------|
| b=8, h=18, d=256, s=2560, group | 12.1 ms (200 TFLOPS) | **3.65 ms (662 TFLOPS)** | **3.3x** |
| b=16, h=18, d=256, s=2560, group | 23.6 ms (205 TFLOPS) | **7.15 ms (676 TFLOPS)** | **3.3x** |

## Accuracy

PyTorch path (test_mha_varlen.py): dQ/dK/dV max diff ~0.004 at b=2 s=256. NRMSE ~2e-4 through PyTorch (Section 7 of independent repro confirms).

C++ bwd.exe reports higher NRMSE (~9e-2) for dQ/dK due to BF16 ODO D values in the C++ reference — this is a reference precision issue, not kernel accuracy. BF16 D is intentional (avoids extra F32 workspace).

## Test plan

Python (`test_mha_varlen.py`) on MI355X:

| Config | Result |
|--------|--------|
| b=2 h=8 s=256 (main benchmark) | dQ/dK/dV max diff 0.004, 58 TFLOPS bwd |
| b=2 h=8 s=256 (seq_padding: mixed, q_only, k_only, no_padding) | All pass (causal falls through to CK) |
| b=8 h=18 s=2560 | 597 TFLOPS bwd, correct |
| b=8 h=18 s=5121 (~41K tokens) | dQ/dK/dV max diff < 0.005, no ODO fault |

C++ (`bwd.exe`) on MI355X:

- [x] Simulator: 9/9 varlen tests pass (EqualSegments, UnequalSegments, ManySmallSegments, PaddedSegments, PaddedEqualSegments, DiffQKLengths)
- [x] GPU (MI355X): dQ/dK/dV max diff 0.004 at b=2 s=256
- [x] GPU (MI355X): bwd.exe 676 TFLOPS at b=16 s=2560, no ODO fault
- [x] Independent repro: PyTorch NRMSE ~2e-4, C++ ~632 TFLOPS at b=8 s=2560
- [x] C++ API (`bwd.exe`): b*h*s*d layout with sq != sk now passes (`valid:y`) — fixed dQ head stride in varlen SRD rebase

### CODE REVIEW

**Review by valarLip — DISMISSED** [review]

**Comment by valarLip** [comment]
@shengnxu  take a look this one？

**Review by slippedJim — CHANGES_REQUESTED** [review]
Hi, we gave a test on this PR and found some issues. You can reproduce with repro_5376.py

[repro_5376.py](https://github.com/user-attachments/files/32188553/repro_5376.py)

[review_5376.md](https://github.com/user-attachments/files/32188563/review_5376.md)

**Comment by alexioslyrakis-amd** [comment]
Hi @slippedJim, thanks for the detailed review! I addressed all points and committed the fixes. See the results.
```
  ┌──────┬──────────────────┬─────────────────────┬─────────────────────────────────────────────────────────────────────────────┐
  │  #   │      Issue       │    GPU Verified?    │                                   Result                                    │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P0-1 │ dQ unwritten     │ Yes (Section A)     │ 0 dropped rows for all sq/sk combos                                         │
  │      │ rows             │                     │                                                                             │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P0-2 │ GQA GPU fault    │ Yes (repro gqa)     │ RuntimeError instead of GPU fault; Python gate falls back to CK             │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P1-1 │ ODO OOB          │ Yes (repro odo)     │ "survived" — no fault                                                       │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P1-2 │ C++ causal guard │ No repro case in    │ Code-reviewed; guard combined with GQA: mt != 0 || nhead_q != nhead_k       │
  │      │                  │ script              │                                                                             │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P1-3 │ fp16 hd256       │ Yes (Section B)     │ "hdim=256 fp16: ok" — falls back to CK                                      │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P1-4 │ hdim 200-248     │ Yes (Section B)     │ "hdim=200/240: ok" — fall back to CK                                        │
  ├──────┼──────────────────┼─────────────────────┼─────────────────────────────────────────────────────────────────────────────┤
  │ P2   │ logits_soft_cap  │ Yes (check script)  │ Gate works — ASM blocked, CK runs. CK also ignores cap (pre-existing CK     │
  │      │                  │                     │ limitation, not our bug)                                                    │
  └──────┴──────────────────┴─────────────────────┴─────────────────────────────────────────────────────────────────────────────┘
``` 

Can you please have another look?

File: csrc/cpp_itfs/mha_bwd.cu [hunk 1]
```diff
@@ -689,8 +698,10 @@ float fmha_v3_bwd(mha_bwd_args a, const ck_tile::stream_config& s)
 
     auto dqdkdv_kernel_launch = [&]() {
         arg_size                  = sizeof(dqdkdv_args);
-        int bdx = (arch_id == "gfx1250") ? 128 : 256;
-        int gdx = (a.max_seqlen_k + ts_kv - 1) / ts_kv;
+        int bdx = (arch_id == "gfx1250") ? 128 : (a.hdim_q >= 256 ? 512 : 256);
+        int gdx_kv = (a.max_seqlen_k + ts_kv - 1) / ts_kv;
+        int gdx_q  = (a.max_seqlen_q + ts_kv - 1) / ts_kv;
+        int gdx = std::max(gdx_kv, gdx_q);
```
Comment (slippedJim) [inline csrc/cpp_itfs/mha_bwd.cu]: This change will lead to accuracy regression to existing asm kernel.

**Review by slippedJim — CHANGES_REQUESTED** [review]

File: csrc/cpp_itfs/mha_bwd.cu (hunk 1 above)
Comment (alexioslyrakis-amd) [reply csrc/cpp_itfs/mha_bwd.cu]: Right, good point! I limited this change only for my kernel. The other kernels must now be unaffected.

File: aiter/ops/mha.py [hunk 2]
```diff
@@ -3213,6 +3211,7 @@ def _flash_attn_varlen_backward(
     is_v3_atomic_fp32: bool | None = True,
     how_v3_bf16_cvt: int | None = 1,
     zero_tensors: bool = False,
+    logits_soft_cap: float = 0.0,
```
Comment (slippedJim) [inline aiter/ops/mha.py]: Why add this?

File: aiter/ops/mha.py [hunk 3]
```diff
@@ -3297,68 +3297,26 @@ def can_impl_fmha_v3_bwd_gfx950():
...
-        # `can_impl_fmha_v3_bwd*` gates exclude by requiring hdim_q == hdim_v.
-        # `deterministic` is absent on purpose: the kernel uses no atomics and
-        # writes each of dq/dk/dv exactly once, so it is deterministic either way.
-        ret = get_gfx() == "gfx942"
-        ret &= alibi_slopes is None
-        ret &= dropout_p == 0.0
-        ret &= hdim_q == 192 and hdim_v == 128
-        ret &= nhead_q == nhead_k
+        ret &= (
+            (hdim_q in (128, 256)) or (hdim_q == 192 and hdim_v == 128)
```
Comment (slippedJim) [inline aiter/ops/mha.py]: (hdim_q == 192 and hdim_v == 128) conflict with L3298: ret &= hdim_q == hdim_v

Comment (slippedJim) [inline aiter/ops/mha.py]: hdim_q in (128, 256) has change the origin range `hdim_q > 64 and hdim_q`. This leads to hdim64 and hdim padding cases regression(fallback to CK).

**Review by slippedJim — CHANGES_REQUESTED** [review]

File: aiter/ops/mha.py (hunk 2 above)
Comment (alexioslyrakis-amd) [reply aiter/ops/mha.py]: YOu're right, removed! the forward gate already rejects logits_soft_cap != 0, so the backward never sees it

File: aiter/ops/mha.py (hunk 3 above)
Comment (alexioslyrakis-amd) [reply aiter/ops/mha.py]: Good catch, removed the dead branch

Comment (alexioslyrakis-amd) [reply aiter/ops/mha.py]: Fixed by restoring the original lower bound hdim_q > 64 and extended the upper to 256.

**Review by slippedJim — APPROVED** [review]
LGTM
